Guide.docc/IncrementalAdoption.md

```diff
@@ -1,3 +1,236 @@
+common pattern in Swift.
+It's possible to make a version of such a function that is usable directly from 
+an asynchronous context.
+
+```swift
+func updateStyle(backgroundColor: ColorComponents, completionHandler: @escaping () -> Void) {
+    // ...
+}
+```
+
+This is an example of a function that informs a client its work is complete
+using a callback.
+You can wrap this function up into an asynchronous version using
+_continuations_.
+
+```swift
+func updateStyle(backgroundColor: ColorComponents) async {
+    withCheckedContinuation { continuation in
+        updateStyle(backgroundColor: backgroundColor) {
+            continuation.resume()
+        }
+    }
+}
+```
+
+The `withCheckedContinuation` function is one of a [suite of standard library
+APIs][continuation-apis] that exist to make interfacing non-async and async code possible.
+
+> Note: Introducing asynchronous code into a project can surface data isolation
+checking violations. To understand and address these, see [Crossing Isolation Boundaries][]
+
+[Crossing Isolation Boundaries]: commonproblems#Crossing-Isolation-Boundaries
+[continuation-apis]: https://developer.apple.com/documentation/swift/concurrency#continuations
+
+## Dynamic Isolation
+
+Expressing the isolation of your program statically, using annotations and
+other language constructs, is both powerful and concise.
+But, because static isolation uses Swift's type system, introducing it
+can potentially require changes to all places a type is used.
+
+Dynamic isolation provides runtime mechanisms for describing data isolation.
+These are a more manual, but also more flexible alternative.
+Dynamic isolation can be an essential tool for interfacing a Swift 6 component
+with another that has not yet migrated,
+even if these components are within the _same_ module.
+
+### Internal-Only Isolation
+
+Suppose you have determined that a reference type within your project can be
+best described with `MainActor` static isolation.
+
+```swift
+@MainActor
+class WindowStyler {
+    private var backgroundColor: ColorComponents
+
+    func applyStyle() {
+        // ...
+    }
+}
+```
+
+This `MainActor` isolation may be _logically_ correct.
+But, if this type is used in other unmigrated locations,
+adding static isolation here could require many additional changes.
+An alternative is to use dynamic isolation to help control the scope.
+
+```swift
+class WindowStyler {
+    @MainActor
+    private var backgroundColor: ColorComponents
+
+    func applyStyle() {
+        MainActor.assumeIsolated {
+            // use and interact with other `MainActor` state
+        }
+    }
+}
+```
+
+Here, the isolation has been internalized into the class.
+This keeps any changes localized to the type, allowing you make
+changes without affecting any clients of the type.
+
+However, a disadvantage of this technique is now the type and its internals
+do not have the same isolation domain.
```

Comment:
I think the main disadvantage of hiding isolation from _clients_ is that you do not get the benefit of concurrency invariants being documented in the public API surface. IMO we should strongly advise against this for public API.

Reply:
yeah you are totally right. Need to make sure this is clearly not a great long-term idea.